File: 03.go
package main

import (
	"bufio"
	"fmt"
	"os"
)

type Vector struct {
	X int
	Y int
}

func hits(grid []string, dir Vector) int {
	var result int
	var pos Vector
	size := Vector{len(grid[0]), len(grid)}
	for pos.Y < size.Y {
		if grid[pos.Y][pos.X%size.X] == '#' {
			result++
		}
		pos.X += dir.X
		pos.Y += dir.Y
	}
	return result
}

func main() {
	var grid []string
	scanner := bufio.NewScanner(os.Stdin)
	for scanner.Scan() {
		grid = append(grid, scanner.Text())
	}

	// part 1
	fmt.Println(hits(grid, Vector{3, 1}))

	// part 2
	dirs := []Vector{
		{1, 1},
		{3, 1},
		{5, 1},
		{7, 1},
		{1, 2},
	}
	product := 1
	for _, d := range dirs {
		product *= hits(grid, d)
	}
	fmt.Println(product)
}

File: 03.py
import fileinput
import math

def hits(grid, dx, dy):
    w = len(grid[0])
    h = len(grid)
    x = y = count = 0
    while y < h:
        if grid[y][x%w] == '#':
            count += 1
        x += dx
        y += dy
    return count

grid = [x.strip() for x in fileinput.input()]

# part 1
print(hits(grid, 3, 1))

# part 2
dirs = [(1, 1), (3, 1), (5, 1), (7, 1), (1, 2)]
counts = [hits(grid, dx, dy) for dx, dy in dirs]
print(math.prod(counts))
